app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationsPagingSource.kt

```diff
@@ -0,0 +1,184 @@
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+import okhttp3.Headers
+import retrofit2.Response
+import javax.inject.Inject
+
+/** Models next/prev links from the "Links" header in an API response */
+data class Links(val next: String?, val prev: String?)
+
+/** [PagingSource] for Mastodon Notifications, identified by the Notification ID */
+class NotificationsPagingSource @Inject constructor(
+    private val mastodonApi: MastodonApi,
+    private val notificationFilter: Set<Notification.Type>
+) : PagingSource<String, Notification>() {
+    override suspend fun load(params: LoadParams<String>): LoadResult<String, Notification> {
+        Log.d(TAG, "load() with ${params.javaClass.simpleName} for key: ${params.key}")
+
+        try {
+            val response = when (params) {
+                is LoadParams.Refresh -> {
+                    getInitialPage(params)
+                }
+                is LoadParams.Append -> mastodonApi.notifications(
+                    maxId = params.key,
+                    limit = params.loadSize,
+                    excludes = notificationFilter
+                )
+                is LoadParams.Prepend -> mastodonApi.notifications(
+                    minId = params.key,
+                    limit = params.loadSize,
+                    excludes = notificationFilter
+                )
+            }
+
+            if (!response.isSuccessful) {
+                return LoadResult.Error(Throwable(response.errorBody().toString()))
+            }
+
+            val links = getPageLinks(response.headers()["link"])
+            return LoadResult.Page(
+                data = response.body()!!,
+                nextKey = links.next,
+                prevKey = links.prev
+            )
+        } catch (e: Exception) {
+            return LoadResult.Error(e)
+        }
+    }
+
+    /**
+     * Fetch the initial page of notifications, using params.key as the ID of the initial
+     * notification to fetch.
+     *
+     * - If there is no key, a page of the most recent notifications is returned
+     * - If the notification exists, and is not filtered, a page of notifications is returned
+     * - If the notification does not exist, or is filtered, the page of notifications immediately
+     *   before is returned
+     * - If there is no page of notifications immediately before then the page immediately after
+     *   is returned
+     */
+    private suspend fun getInitialPage(params: LoadParams<String>): Response<List<Notification>> = coroutineScope {
+        // If the key is null this is straightforward, just return the most recent notifications.
+        val key = params.key
+            ?: return@coroutineScope mastodonApi.notifications(
+                limit = params.loadSize,
+                excludes = notificationFilter
+            )
+
+        // It's important to return *something* from this state. If an empty page is returned
```

Comment:
This method sounds a bit complicated?

(So complicated that I have no idea how to simplify it, but it should...)

Maybe the contract should be much clearer: not 4 different more or less distinct possibilities?

Reply:
Blame the Mastodon API, it's poorly designed here.

I think the method itself is quite simple, but the design has to be quite careful. As noted in the Matrix chat we think Tusky is failing to fetch some post in some cases. The design here is to avoid the corner cases where that can happen, and the extensive commenting is so that someone doesn't come along in the future and try and simplify it without properly understanding the corner cases.